Updating product AI features endpoints

joelclimbsthings opened the pull request:
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Switching the product title generation and description generation features over to using our generic text completion endpoint. 

Closes #38895 

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout branch and activate plugin.
2. Navigate to WooCommerce -> Add new
3. Click on title field, and generate suggestions using box that appears below. This behavior should function as it has previously.
4. Click the "Write with AI" button above the description field. This should stream to the contents of the field. It's behavior is also the same as it was previously.

<!-- End testing instructions -->

joelclimbsthings committed:
Switching front-end description logic to use new endpoint

joelclimbsthings committed:
Logging troubles

joelclimbsthings commented on plugins/woo-ai/api/api.php:
```diff
@@ -17,7 +17,8 @@
  * Register the Woo AI route.
  */
 function woo_ai_rest_api_init() {
+	error_log( 'this one is not' );
```
As noted here, for some reason none of my `error_log`'s are showing up for the API logic.

joelclimbsthings commented:
@nima-karimi @tommyshellberg I have the description logic working with the new endpoint easily enough, although I'm having trouble migrating the title feature. 

I'm receiving this error from the request:
```
{
    "code": "error",
    "message": "Failed to fetch the suggestions: Failed to get completion ( reason: Bad Request )",
    "data": {
        "status": 400
    }
}
```

So I believe [this line](https://github.com/woocommerce/woocommerce/blob/abbebdcf764d9bcc323a3d99b46b821ef6f07411/plugins/woo-ai/includes/product-data-suggestion/class-product-data-suggestion-service.php#L70) is being hit, but for some reason none of my logging is working in the API files. I'm fairly unfamiliar with the `wpcom_json_api_request_as_user` function, so the lack of logging is an issue.

The main issue is my inability to log things (see my comment on the code change). Any idea why that would be the case? It works until all of code loaded with the `rest_api_init` hook.

joelclimbsthings committed:
Tentively working for title suggestions

joelclimbsthings committed:
Adding changelog

joelclimbsthings commented on plugins/woo-ai/src/hooks/useProductDataSuggestions.ts:
```diff
@@ -24,14 +25,15 @@ export const useProductDataSuggestions = () => {
+			);
 
-			return response.suggestions;
+			return JSON.parse( response.completion ).suggestions;
```
This JSON parsing should likely be done server-side, but I was having trouble getting this done since my logging isn't working within that code.

joelclimbsthings committed:
Updating to accept altered response body for title suggestions

joelclimbsthings committed:
Using client-size streaming instead of our own proxy API

joelclimbsthings committed:
Removing obsolete API code

tommyshellberg commented on plugins/woo-ai/src/hooks/useProductDataSuggestions.ts:
```diff
@@ -25,15 +25,16 @@ export const useProductDataSuggestions = () => {
-					method: 'POST',
-					data: { ...request, token },
-				} )
+			const { suggestions } = await requestJetpackToken().then(
```
Perhaps nitpicky but what about a more consistent approach to Promises (either using async/await or promise chains but not both)?
I like async/await because it's more readable:

```
  const { token } = await requestJetpackToken();
  
   const { suggestions }  = await apiFetch<ProductDataSuggestionSuccessResponse>({
    path: '/wooai/product-data-suggestions',
    method: 'POST',
    data: { ...request, token },
  });
```

joelclimbsthings replied:
Interesting thought, I actually like to use both at times depending on the circumstances. I believe it might be because otherwise you end up with a bunch of temporary variables that you don't necessarily need. Also, if you have a longer chain it actually seems more readable to me that way, although with just the two requests above I can see it feeling less readable.

```js
const finalResult = await doThing().then(doAnotherThing)
			.then(doYetAnotherThing)
			.then(doTheLastThing)
			.catch(e -> handleError);
```

That said, glad to change it in this case 👍🏻

tommyshellberg replied:
> Also, if you have a longer chain it actually seems more readable to be that way, although with just the two requests above I can see it feeling less readable.
> 
> ```js
> const finalResult = await doThing().then(doAnotherThing)
> 			.then(doYetAnotherThing)
> 			.then(doTheLastThing)
> 			.catch(e -> handleError);
> ```

That can be true for sure, and I do love the concept of method chaining. 😆 
Although, the counter-argument is that it makes debugging quite a lot more difficult, right?
And, while it is super practical to put a `.catch()` at the end to catch everything, you can accomplish the same with `try{} catch{}` blocks. 🤔 

And just as an FYI, from time to time I have recommendations/questions like this which aren't really meant as criticisms but rather serve as learning experiences for me, given that I don't have a _ton_ of experience, so it's helpful for me to hear different opinions.  🙇

joelclimbsthings replied:
> And just as an FYI, from time to time I have recommendations/questions like this which aren't really meant as criticisms but rather serve as learning experiences 

For sure! They're potential learning experiences for both of us, and it's always good to question those things that I'm just "in the habit" of doing (and not always certain why). Keep'm coming!

tommyshellberg commented on plugins/woo-ai/src/utils/productData.ts:
```diff
@@ -36,6 +37,7 @@ const getTags = (): string[] => {
 	return tags.filter( ( tag ) => tag !== '' );
 };
 
+// TODO: Not working in my testing when manually adding attributes.
```
Can we just delete this, then?

joelclimbsthings replied:
I _believe_ these may work in other contexts, such as if you're editing an existing product, so we may just want to improve them. @nima-karimi could clarify more here.

nima-karimi replied:
hmm. Could you share your test steps? If you press the "Save Attributes" button then it should include those attributes in the request 🤔

joelclimbsthings replied:
Thanks @nima-karimi , it looks like the categories selectors only worked when those categories had been applied on the last page refresh, and not recently added/checked. 

For the attributes, it looked like it was only working for global attributes, and not local I believe. I've fixed those both in the last couple commits ([1](https://github.com/woocommerce/woocommerce/pull/38930/commits/81f15148873ee4443dc8d6e4d0481317c2e63cff), [2](https://github.com/woocommerce/woocommerce/pull/38930/commits/e370aba1d52c26791cf9ee0dc0951840a3c2ef08))

tommyshellberg commented on plugins/woo-ai/src/utils/productData.ts:
```diff
@@ -7,6 +7,7 @@ import { getTinyContent, getPostId } from '.';
 const isElementVisible = ( element: HTMLElement ) =>
 	! ( window.getComputedStyle( element ).display === 'none' );
 
+// TODO: This appears to be returning category IDs instead of names?
```
If it's not working properly, I'm in favor of just removing it and re-adding it later when we can properly handle it.

joelclimbsthings replied:
See my [other](https://github.com/woocommerce/woocommerce/pull/38930#discussion_r1251150706) comment.

tommyshellberg commented:
This is working for me and it's great that both titles and descriptions are being generated through the same endpoint, nice work @joelclimbsthings 🥳 

One thing I noticed is that the new prompt will send quite a lot of product data that doesn't seem like it would help an LLM write a description or title for a product:

- product ID (an integer)
- published status ('draft', 'published')

Additionally, there are comments that attributes and categories aren't actually being retrieved properly right now, should we just remove them from productData and add them when they can be used properly?

![Screen Shot on 2023-06-29 at 12:31:24](https://github.com/woocommerce/woocommerce/assets/6723003/a695160f-9e81-4185-953b-ef6f9e8d8feb)

tommyshellberg commented on plugins/woo-ai/src/product-name/product-name-suggestions.tsx:
```diff
@@ -56,8 +53,35 @@ export const ProductNameSuggestions = () => {
+				setSuggestions( parsed.suggestions );
+				setIsFirstLoad( false );
+			} catch ( e ) {
+				throw new Error( 'Unable to parse suggetsions' );
```
Slight typo:
```suggestion
				throw new Error( 'Unable to parse suggestions.' );
```

tommyshellberg commented on plugins/woo-ai/src/product-name/product-name-suggestions.tsx:
```diff
@@ -56,8 +53,35 @@ export const ProductNameSuggestions = () => {
+			} );
+			setSuggestionsState( SuggestionsState.Failed );
+		},
+		onCompletionFinished: ( reason, content ) => {
```
I did not expect that the product titles would use the streaming endpoint but it works because of this `onCompletionFinished()` function, nice thinking here!
I think I like the idea of streaming _all the things_ which then adds the possibility of rendering title suggestions one at a time as they come in, further reducing the perception of latency.
We would have to build a function that could parse the JSON string being returned from the EventSource and pick out valid product title objects, and when we find them, add them via `setSuggestions()`. 
Do you think it's worth adding as an enhancement for later?

joelclimbsthings replied:
Definitely think this could be a worthy enhancement 👍

tommyshellberg commented:
Overall, I like this approach and I'm really happy about this:
![Screen Shot 2023-06-29 at 12 25 27](https://github.com/woocommerce/woocommerce/assets/6723003/9b99c9d3-b7ee-4b4b-a255-5281ab9c3ffa)
This makes the code more unified and easier to understand going forward. 💯

joelclimbsthings committed:
implementing feedback

joelclimbsthings committed:
Refactoring product data logic

joelclimbsthings committed:
Fixing getAttributes function to work with local attributes

joelclimbsthings commented:
Thanks for the reviews @nima-karimi and @tommyshellberg . I've reduced the amount of product data relayed in the title suggestions prompt, and repaired the attributes and categories functions. I believe I've addressed the feedback; let me know!

joelclimbsthings committed:
Allowing for boolean values in product data

nima-karimi commented on plugins/woo-ai/src/utils/productData.ts:
```diff
@@ -1,32 +1,25 @@
  */
-import { Attribute, ProductData } from './types';
-import { getTinyContent, getPostId } from '.';
+import { Attribute } from './types';
```
Let's also remove the type definition from the ./types.

joelclimbsthings replied:
Not sure I understand. We're still using this type, so I'm not sure why we want to delete it?

nima-karimi replied:
Sorry I meant the ProductData type

joelclimbsthings replied:
Right! Thanks 👍🏻

nima-karimi commented on plugins/woo-ai/src/utils/productData.ts:
```diff
@@ -1,32 +1,25 @@
-
-const getCategories = (): string[] => {
-	const categoryCheckboxEls: NodeListOf< HTMLInputElement > =
+export const getCategories = () =>
```
If we can change this always to return an array of strings (and not undefined), it would be more predictable. It can return an empty array if nothing is found. Something like this:

```typescript
export const getCategories = (): string[] => {
	const categoryInputs: NodeListOf< HTMLInputElement > =
		document.querySelectorAll< HTMLInputElement >(
			'#taxonomy-product_cat input[name="tax_input[product_cat][]"]'
		);

	if ( ! categoryInputs.length ) {
		// if no results, return an empty array
		return [];
	}

	return Array.from( categoryInputs )
		.filter(
			( inputElement: HTMLInputElement ) =>
				window.getComputedStyle( inputElement, ':before' ).content !==
				'none'
		)
		.map(
			( inputElement: HTMLInputElement ) =>
				inputElement.nextSibling?.nodeValue?.trim() ?? ''
		);
};
```

joelclimbsthings replied:
Good call! Done in 421f3534539d5bc3f2bde5aaabc9eb51e4c91ad8

nima-karimi commented on plugins/woo-ai/src/product-name/product-name-suggestions.tsx:
```diff
@@ -177,42 +206,61 @@ export const ProductNameSuggestions = () => {
+			'Product titles should contain at least 20 characters.',
+			'Return a short and concise reason for each suggestion in seven words in the "reason" part of your response.',
+			"The product's properties are:",
+			`${ JSON.stringify( validProductData ) }`,
```
Instead of passing JSON in the prompt, could we pass a comma-separated "key: value" list? It would also be helpful to convert booleans to Yes/No values and remove any underlines from the keys. It might save us a few tokens and be more similar to natural language.

joelclimbsthings replied:
Right, we could remove the brackets, although it would add in several newline characters, so I'm not sure it would save us any tokens. I had the thought that ChatGPT may actually like this to be more "structured," and it cleanly delineates it from the rest of the prompt. That said, I'm a little ambivalent, although it would make the code a bit more complex 🤔

nima-karimi reviewed (commented):
Thanks for updating this, @joelclimbsthings! 

The code looks good, but I still haven't tested it. Since we are now sending additional data to LLM (product type, is downloadable, and is virtual), it would be good to run a few tests and see the response. I'll work on setting up a WoA site to test and approve it. 

In the meantime, I noticed a few minor things and left some comments. Let me know what you think.

joelclimbsthings committed:
Refactoring getCategories to always return array

joelclimbsthings commented:
Thanks for the review @nima-karimi ! I've implemented your feedback, and responded to a couple comments. Let me know what you think!

joelclimbsthings committed:
Removing orphaned type

joelclimbsthings committed:
Refactoring getCategories again

nima-karimi commented on plugins/woo-ai/src/utils/text-completion.ts:
```diff
@@ -88,20 +84,16 @@ async function requestToken() {
  * @param {string} question - The query to send to the API
  * @param {number} postId   - The post where this completion is being requested, if available
```
Missed deleting this one 😅

joelclimbsthings replied:
Whoops! 👍🏻

nima-karimi commented:
> I have noticed that when you give it data, it tends to want to use it, which may not always be ideal. 

Exactly, I noticed this, too, when working with the descriptions. We should modify the prompt to put less weight on the properties.

nima-karimi reviewed (approved):
This tests fine. I think the prompt could definitely use some improvement because of the issue you mentioned above, but that could be a separate PR. I left one tiny comment but otherwise good to go :shipit:

joelclimbsthings committed:
Cleanup of text completion params

joelclimbsthings committed:
Small prompt tweaks
